ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,561 @@
+            "mcast_flood_unregistered": "false",
+            "mcast_snoop": "false",
+            "vlan-passthru": "false",
+        }
+
+        neutron_net_name = f"neutron-{uuid.uuid4()}"
+        neutron_network = self.nbctl.ls_add(
+            neutron_net_name,
+            self.next_int_net(),
+            ext_ids=switch_ext_ids,
+            other_config=switch_config,
+        )
+
+        return neutron_network
+
+    def _add_network_subnet(self, network: NeutronNetwork) -> None:
+        """Create DHCP subnet for a network.
+
+        Adding subnet to a network in Neutron results in DHCP_Options being
+        created in OVN.
+        """
+        external_ids = {
+            "neutron:revision_number": "0",
+            "subnet_id": str(uuid.uuid4()),
+        }
+        static_routes = (
+            f"{{169.254.169.254/32,{network.gateway}, "
+            f"0.0.0.0/0,{network.gateway}}}"
+        )
+        options = {
+            "classless_static_route": static_routes,
+            "dns_server": "{1.1.1.1}",
+            "lease_time": "43200",
+            "mtu": "1442",
+            "router": str(network.gateway),
+            "server_id": str(network.gateway),
+            "server_mac": str(RandMac()),
+        }
+        dhcp_options = self.nbctl.create_dhcp_options(
+            str(network.network.cidr), ext_ids=external_ids
+        )
+
+        self.nbctl.dhcp_options_set_options(dhcp_options.uuid, options)
+
+    def _assign_port_ips(self, network: NeutronNetwork) -> None:
+        """Assign IPs to each LS port associated with the Network."""
+        for uuid_, port in network.ports.items():
+            if port.ip is None or port.ip == "unknown":
+                updated_port = self.nbctl.ls_port_set_ipv4_address(
+                    port, str(network.next_host_ip())
+                )
+                network.ports[uuid_] = updated_port
+                # XXX: Unable to update external_ids
+
+    def _add_network_port(self, network: LSwitch, project_id: str) -> LSPort:
+        """Create regular port in LSwitch with Neutron's external IDs."""
+        port_ext_ids = {
+            "neutron:cidrs": "",
+            "neutron:device_id": f"ovnmeta-{network.uuid}",
+            "neutron:device_owner": "network:distributed",
+            "neutron:network_name": network.name,
+            "neutron:port_name": "",
+            "neutron:project_id": project_id,
+            "neutron:revision_number": "1",
+            "neutron:security_group_ids": "",
+            "neutron:subnet_pool_addr_scope4": "",
+            "neutron:subnet_pool_addr_scope6": "",
+        }
+
+        port = self.nbctl.ls_port_add(
+            lswitch=network,
+            name=str(uuid.uuid4()),
+            mac=str(RandMac()),
+            ext_ids=port_ext_ids,
+        )
+        self.nbctl.ls_port_set_set_type(port, "localport")
+        self.nbctl.ls_port_set_set_options(port, "requested-chassis=")
+        self.nbctl.ls_port_enable(port)
```

Comment:
This is a general remark. It looks like you are not using transactions, whereas the ML2 driver tries to do so as much as possible. I'm not sure if it has any performance implications. If not, maybe that's not important.

Reply:
Thanks for the tip about the transaction, I did not consider them. I assume that we would not be measuring/profiling the "cluster_bringup" but better performance never hurts. I'll look into where I could group calls into transactions :+1:

Reply:
I've looked into this bit more today and I don't see a nice way to integrate transactions. Currently I'm using mainly `ovn_utils.OvnNbctl` moethods which are not "batch-able". I could get around it by using `OvnNbctl.idl` directly but then we'd lose code reusing.
Also, correct me if I'm wrong here, but ovn-heater does not measure/profile "cluster_bringup", does it? If not, I'd lean towards not implementing transactions now.

Reply:
I don't know if cluster bringup is measured by ovn-heater. But that's ok if we don't use transactions. At worst we'll get lower results. Since CMS's most likely use them, performance can only improve :)